keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +337,302 @@ or later. The dual-stack feature requires no changes to this API.
+The below are sample services that demonstrate the above rules
+
+```yaml
+# dual stack service (required)
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  preferDualStack: true # set by user
+  ipFamilies: # set by user
+    - IPv4
+    - IPv6
+  clusterIP: 1.2.3.4  # allocated
+  clusterIPs: # allocated
+    - 1.2.3.4
+    - 2001::1
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+
+# dual stack service (required)
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  preferDualStack: true # defaulted
+  ipFamilies: # defaulted
+    - IPv6
+    - IPv4
+  clusterIP: 2001::1  # set by user
+  clusterIPs: # set by user
+    - 2001::1
+    - 1.2.3.4
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+
+# dual stack service (optional)
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  preferDualStack: true # set by user 
+  ipFamilies:
+    - IPv6 # defaulted, cluster does not support dual stack services, only one ipv6 cidr is provided for services
+  clusterIP: 2001::1  # allocated
+  clusterIPs: # allocated
+    - 2001::1 # note single entry
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+
+# dual stack service (optional)
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  preferDualStack: true # set by user 
+  ipFamilies:
+    - IPv6 # set by user
+    - IPv4 # defaulted (cluster supports dual stack)
+  clusterIP: 2001::1  # allocated
+  clusterIPs: # allocated
+    - 2001::1
+    - 1.2.3.4
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+
+```
+
+##### Notes on NodePort Allocation
+
+NodePort reservation will be constant across both families. That mean a reservation of `NodePort: 12345` will happen across both families irrespective of the ip family of the service.
 
 #### Type Headless services
 
-\<TBD\>
+1. For Headless with selector, the value of `spec.preferDualStack` and `spec.ipFamilies` will drive the endpoint selection. Endpoint slice controller will select endpoints with ip families that match the value of these two flags.
+2. For Headless without selector the values of `spec.preferDualStack` and `spec.ipFamilies` are treated as saved as they were provided by the user.
+
+> in both cases the values of `spec.preferDualStack` and `spec.ipFamilies` are validated semantically. 
 
 #### Type ExternalName
 
-\<TBD\>
+For ExternalName service the value of `spec.ipFamilyPolicy` and `spec.ipFamilies` is expected to be `nil`. Values will be set to `nil` by the api-server.
+
+#### Services and Pod Environment Variables
+Kubernetes injects environment variables into each pod representing services that are in the same namespace. Existing variables will not change. The below will be added
```

Comment:
Yeah, I REALLY don't think we should do this - env vars are already a burden and we want to not make that worse.  Convince me otherwise?

Reply:
While they are a burden. They are the only way clientgo can get the default service IP without having to do a call to api-server. I am Ok if we want to remove the entire env var.

thoughts?

Reply:
Down with environment variables!

Reply:
As above - let's leave them with the [0] address only and see if anyone screams?

Reply:
The argument i have had in mind was a for a pod working on secondary networking stack (e.g. `ipv4` on a primary `ipv6` cluster to be able to access api server). But that will also work because the two network stacks will be configured within pod networking. 

Unless something is really funky (e.g. raw sockets or the likes forcing `ipv4` only to `defaultService.ClusterIP` ) with pod/container code then it should work as expected.

will remove .. yay less work for me..